### PR TITLE
Prevent clippy::needless_lifetimes false positive in async function definition

### LINKED ISSUE #7893: Lifetime cannot be elided
Lint name: clippy::needless_lifetimes

I tried this code:

```rust
async fn func<'a>(args: &[&'a str]) -> Option<&'a str> {
    args.get(0).cloned()
}
```

I expected to see this happen: no warning as the `'a` lifetime is needed. Moreover the lint does not trigger is the function is not marked as `async`.

Instead, this happened: 

```
warning: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
 --> src/main.rs:1:1
  |
1 | async fn func<'a>(args: &[&'a str]) -> Option<&'a str> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::needless_lifetimes)]` on by default
```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.58.0-nightly (4e0d3973f 2021-10-27)
binary: rustc
commit-hash: 4e0d3973fafdfb1c51011bc74e44257b5e3863f1
commit-date: 2021-10-27
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```

### PR DESCRIPTION
**Pull request by samueltardieu**
Scan `OpaqueDef` bounds for lifetimes as well. Those `OpaqueDef` instances are generated while desugaring an `async` function definition.

This fixes #7893

changelog: Prevent [`clippy::needless_lifetimes`] false positive in `async` function definition

### CODE REVIEW

**Review by xFrednet — APPROVED** [review]
This looks good to me :+1:. I'm not quite sure how these `OpaqueDef` types are used inside rustc, I'll therefore ask some team members to double-check this PR :upside_down_face:

**Comment by samueltardieu** [comment]
> I'm not quite sure how these `OpaqueDef` types are used inside rustc,

According to [the documentation of `OpaqueDef`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_hir/enum.TyKind.html#variant.OpaqueDef), this is the type used to expand `impl Trait` (and `async fn` gets expanded into a function returning `impl Future<_>`). The documentation also indicates that the second field of `OpaqueDef` contains the lifetimes bound on the `impl Trait`, hence this patch.

File: tests/ui/needless_lifetimes.rs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+// edition:2018
```
Comment (flip1995) [inline tests/ui/needless_lifetimes.rs:1]: This shouldn't be necessary anymore, since we use the 2021 edition by default for our tests now.

**Review by flip1995 — APPROVED** [review]
Thanks for the further explanation. This LGTM. There's a small NIT with the test.

File: tests/ui/needless_lifetimes.rs (hunk 1 above)
Comment (samueltardieu) [reply tests/ui/needless_lifetimes.rs:1]: Do we? If I remove it I get an error: "Rust 2015 does not permit the use of `async fn`", and the 3 commits that are not in my "master" branch don't seem to change that.

Comment (flip1995) [reply tests/ui/needless_lifetimes.rs:1]: Huh? Let me actually try out my suggestion myself. It _should_ work after rebasing.

Comment (samueltardieu) [reply tests/ui/needless_lifetimes.rs:1]: I was out of sync with upstream, I'll rebase and update the test!

Comment (flip1995) [reply tests/ui/needless_lifetimes.rs:1]: Thanks! I rebased on the latest upstream master and it worked for me locally.

Comment (samueltardieu) [reply tests/ui/needless_lifetimes.rs]: You were right, rebasing allowed me to remove this line.

**Comment by flip1995** [comment]
@bors r=xFrednet,flip1995

Thanks!

**Pull request merged**